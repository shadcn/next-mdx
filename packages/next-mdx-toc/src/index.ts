import remark from "remark"
import toc from "mdast-util-toc"
import visit from "unist-util-visit"
import { MdxNode } from "next-mdx"

interface Item {
  title: string
  url: string
  items?: Item[]
}

interface Items {
  items?: Item[]
}

function getItems(node, current): Items {
  if (!node) {
    return {}
  }

  if (node.type === "paragraph") {
    visit(node, (item) => {
      if (item.type === "link") {
        current.url = item.url
      }

      if (item.type === "text") {
        current.title = item.value
      }
    })

    return current
  }

  if (node.type === "list") {
    current.items = node.children.map((i) => getItems(i, {}))

    return current
  } else if (node.type === "listItem") {
    const heading = getItems(node.children[0], {})

    if (node.children.length > 1) {
      getItems(node.children[1], heading)
    }

    return heading
  }

  return {}
}

const getToc = () => (node, file) => {
  const table = toc(node)
  file.data = getItems(table.map, {})
}

export interface TableOfContents extends Items {}

export async function getTableOfContents(
  node: MdxNode
): Promise<TableOfContents> {
  const result = await remark().use(getToc).process(node.content)

  return result.data
}
