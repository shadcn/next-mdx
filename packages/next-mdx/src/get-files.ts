import glob from "fast-glob"
import path from "path"

import { getSourceConfig } from "./get-config"

export interface MdxFile {
  filepath: string
  slug: string
  url?: string
}

export async function getFiles(sourceName: string): Promise<MdxFile[]> {
  const { contentPath, basePath } = await getSourceConfig(sourceName)
  const files = await glob(`${contentPath}/**/*.{md,mdx}`)

  if (!files.length) return []

  return files.map((filepath) => {
    let slug = filepath
      .replace(contentPath, "")
      .replace(/^\/+/, "")
      .replace(new RegExp(path.extname(filepath) + "$"), "")

    slug = slug.replace(/\/?index$/, "")

    return {
      filepath,
      slug,
      url: basePath ? `${basePath.replace(/\/$/, "")}/${slug}` : null,
    }
  })
}
