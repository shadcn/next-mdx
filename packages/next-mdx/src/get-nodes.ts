import { promises as fs } from "fs"
import matter from "gray-matter"
import hasha from "hasha"
import { GetStaticPropsContext } from "next"
import { Pluggable, Compiler } from "unified"
import renderToString from "next-mdx-remote/render-to-string"
import { MdxRemote } from "next-mdx-remote/types"

import { mdxCache } from "./get-cache"
import { getFiles, MdxFile } from "./get-files"
import { getConfig, getSourceConfig } from "./get-config"

export type NodeFrontMatter = Record<string, unknown>

// TODO: Properly type node relationships with generics.
export interface NodeRelationships<T = Node> {
  [key: string]: T[]
}

export interface Node<T = NodeFrontMatter> extends MdxFile, MdxFileData<T> {
  mdx: MdxRemote.Source
  relationships?: NodeRelationships
}

// type MdxNodeWithoutMdx<T extends Node> = Omit<T, "mdx">

export interface MdxNode<T = NodeFrontMatter> extends Node<T> {}

export interface MdxParams {
  components?: MdxRemote.Components
  scope?: Record<string, unknown>
  provider?: MdxRemote.Provider
  mdxOptions?: {
    remarkPlugins?: Pluggable[]
    rehypePlugins?: Pluggable[]
    hastPlugins?: Pluggable[]
    compilers?: Compiler[]
    filepath?: string
  }
}

export interface getAllMdxNodesParams extends MdxParams {
  sortBy?: string
  sortOrder?: "asc" | "desc"
}

export interface MdxFileData<T = NodeFrontMatter> {
  hash: string
  frontMatter?: T
  content?: string
}

export async function getMdxNode<T extends MdxNode>(
  sourceName: string,
  context: string | GetStaticPropsContext<NodeJS.Dict<string[]>>,
  params?: MdxParams
): Promise<T> {
  if (!context || (typeof context !== "string" && !context.params?.slug)) {
    new Error(`slug params missing from context`)
  }

  const node = await getNode(sourceName, context)

  if (!node) return null

  return <T>{
    ...node,
    mdx: await renderNodeMdx(node, params),
  }
}

export async function getAllMdxNodes<T extends MdxNode>(
  sourceName: string,
  params?: getAllMdxNodesParams
): Promise<T[]> {
  const nodes = await getAllNodes(sourceName)

  if (!nodes.length) return []

  return Promise.all<T>(
    nodes.map(
      async (node) =>
        <T>{
          ...node,
          mdx: await renderNodeMdx(node, params),
        }
    )
  )
}

async function renderNodeMdx(node: Node, params?: MdxParams) {
  return await renderToString(node.content, {
    ...params,
    scope: {
      ...params?.scope,
      ...node.frontMatter,
    },
  })
}

export async function getNode<T extends Node>(
  sourceName: string,
  context: string | GetStaticPropsContext<NodeJS.Dict<string[]>>
): Promise<T> {
  const files = await getFiles(sourceName)

  if (!files.length) return null

  const slug =
    typeof context === "string"
      ? context
      : context.params.slug
      ? context.params.slug.join("/")
      : ""

  const [file] = files.filter((file) => file.slug === slug)

  if (!file) return null

  const node = await buildNodeFromFile(file)

  return <T>{
    ...node,
    relationships: await getNodeRelationships(node),
  }
}

export async function getAllNodes<T extends Node>(
  sourceName: string
): Promise<T[]> {
  const { sortBy, sortOrder } = await getSourceConfig(sourceName)

  const files = await getFiles(sourceName)

  if (!files.length) return []

  const nodes = await Promise.all<T>(
    files.map(async (file) => {
      const node = await buildNodeFromFile(file)

      return <T>{
        ...node,
        relationships: await getNodeRelationships(node),
      }
    })
  )

  const adjust = sortOrder === "desc" ? -1 : 1
  return <T[]>nodes.sort((a, b) => {
    if (a.frontMatter[sortBy] < b.frontMatter[sortBy]) {
      return -1 * adjust
    }
    if (a.frontMatter[sortBy] > b.frontMatter[sortBy]) {
      return 1 * adjust
    }
    return 0
  })
}

async function buildNodeFromFile<T extends Node>(file: MdxFile): Promise<T> {
  return <T>{
    ...file,
    ...(await getFileData(file)),
    mdx: {
      compiledSource: "",
      renderedOutput: "",
    },
  }
}

export async function getFileData(file: MdxFile): Promise<MdxFileData> {
  const raw = await fs.readFile(file.filepath, "utf-8")
  const hash = hasha(raw.toString())

  const cachedContent = mdxCache.get<MdxFileData>(hash)
  if (cachedContent?.hash === hash) {
    // console.info(`HIT for ${file.slug}`)
    return cachedContent
  }

  // console.info(`MISS for ${file.slug}`)

  const { content, data: frontMatter } = matter(raw)

  const fileData: MdxFileData = {
    hash,
    content,
    frontMatter,
  }

  mdxCache.set<MdxFileData>(hash, fileData)

  return fileData
}

async function getNodeRelationships(node: Node): Promise<NodeRelationships> {
  const relationships: NodeRelationships = {}
  const config = await getConfig()

  for (const key of Object.keys(node.frontMatter)) {
    if (!config[key]) continue

    const values = node.frontMatter[key]

    if (!values) continue

    const valueAsArray: string[] = Array.isArray(values) ? values : [values]
    relationships[key] = await Promise.all(
      valueAsArray.map(async (value) => await getNode(key, value))
    )
  }

  return relationships
}
