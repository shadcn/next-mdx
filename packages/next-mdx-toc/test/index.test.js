import fs from "fs"
import path from "path"
import { getTableOfContents } from "../src"

const doc = fs.readFileSync(
  path.join(__dirname, "./__fixtures__/example.mdx"),
  "utf8"
)

const node = {
  content: doc,
}

test("returns toc tree from node", async () => {
  expect(await getTableOfContents(node)).toMatchSnapshot()
})
