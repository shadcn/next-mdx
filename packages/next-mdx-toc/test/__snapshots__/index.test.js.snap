// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns toc tree from node 1`] = `
Object {
  "items": Array [
    Object {
      "items": Array [
        Object {
          "items": Array [
            Object {
              "title": "Heading 4",
              "url": "#heading-4",
            },
          ],
          "title": "Heading 3",
          "url": "#heading-3",
        },
      ],
      "title": "Heading 2",
      "url": "#heading-2",
    },
    Object {
      "items": Array [
        Object {
          "items": Array [
            Object {
              "title": "Heading 4",
              "url": "#heading-4-1",
            },
            Object {
              "title": "Heading 4",
              "url": "#heading-4-2",
            },
          ],
          "title": "Another heading 3",
          "url": "#another-heading-3",
        },
      ],
      "title": "Another heading 2",
      "url": "#another-heading-2",
    },
  ],
}
`;
